Source/WebKit/UIProcess/mac/WebPageProxyMac.mm

```diff
@@ -765,30 +765,30 @@ static inline bool expectsLegacyImplicitRubberBandControl()
         pageClient->updatePDFHUDLocation(identifier, rect);
 }
 
-void WebPageProxy::pdfZoomIn(PDFPluginIdentifier identifier)
+void WebPageProxy::pdfZoomIn(PDFPluginIdentifier identifier, WebCore::FrameIdentifier frameID)
 {
-    protectedLegacyMainFrameProcess()->send(Messages::WebPage::ZoomPDFIn(identifier), webPageIDInMainFrameProcess());
+    sendToProcessContainingFrame(frameID, Messages::WebPage::ZoomPDFIn(identifier));
 }
 
-void WebPageProxy::pdfZoomOut(PDFPluginIdentifier identifier)
+void WebPageProxy::pdfZoomOut(PDFPluginIdentifier identifier, WebCore::FrameIdentifier frameID)
 {
-    protectedLegacyMainFrameProcess()->send(Messages::WebPage::ZoomPDFOut(identifier), webPageIDInMainFrameProcess());
+    sendToProcessContainingFrame(frameID, Messages::WebPage::ZoomPDFOut(identifier));
 }
 
-void WebPageProxy::pdfSaveToPDF(PDFPluginIdentifier identifier)
+void WebPageProxy::pdfSaveToPDF(PDFPluginIdentifier identifier, WebCore::FrameIdentifier frameID)
 {
-    protectedLegacyMainFrameProcess()->sendWithAsyncReply(Messages::WebPage::SavePDF(identifier), [this, protectedThis = Ref { *this }] (String&& suggestedFilename, URL&& originatingURL, std::span<const uint8_t> dataReference) {
+    sendWithAsyncReplyToProcessContainingFrame(frameID, Messages::WebPage::SavePDF(identifier), Messages::WebPage::SavePDF::Reply { [this, protectedThis = Ref { *this }] (String&& suggestedFilename, URL&& originatingURL, std::span<const uint8_t> dataReference) {
```

Comment:
Can this not just use the completion handler? i.e.

```
std::optional<IPC::AsyncReplyID> WebPageProxy::sendWithAsyncReplyToProcessContainingFrame(std::optional<FrameIdentifier> frameID, M&& message, C&& completionHandler, OptionSet<IPC::SendOption> options)
```

?

Reply:
(as in, do we need to explicitly construct the `Reply`?

Reply:
We do need to explicitly construct the reply because otherwise it tries to instantiate the template with C being the lambda type, which can't be explicitly instantiated in a different translation units because the types of lambdas are special.

Reply:
Makes sense, thanks for the clarification!